### ISSUE TITLE
[Ch 20][Notebook][Major] 사전학습 전/후 [MASK] top-5 서술이 실제 출력과 뒤바뀜 (+ reference BERT 예시 오류)

### ISSUE THREAD

**Issue by fluentmin** [yoon-gu/neuqes-101#60]
## 대상

- **챕터**: Ch 20 — 작은 BERT 직접 사전학습 (영어 MLM, scratch)
- **산출물**: Notebook (`20_en_bert_pretrain/20_en_bert_pretrain.ipynb` — "🔬 학습 직전 baseline" 마크다운 셀, §6-3 도입부, §6-4 뒤 "해석 가이드" 셀)
- **기준 커밋**: `eed3dbe` (챕터 노트북 최종 변경 `8056dab`)
- **환경**: Colab T4 전체 재실행 2회 + 레포 `executed/20_en_bert_pretrain.ipynb` 대조. 두 실행 모두 **에러 0건 / 약 82초**로 정상 완주 — 실행 문제가 아니라 **본문 서술이 실제 출력과 어긋나는** 건입니다.

같은 "해석 가이드" 셀에서 비롯된 두 건이라 하나로 묶어 올립니다. 분리 등록이 필요하면 나눠 드리겠습니다.

### (a) 학습 *전* / 학습 *후* `[MASK]` top-5 서술이 서로 뒤바뀜 — Major

본문은 학습 전(random init)이 **고빈도 기능어**를, 학습 후가 **문맥적 어휘**를 낸다고 서술합니다. 실제 출력은 정확히 그 반대입니다.

| | 본문 주장 | 실제 출력 (2회 실행 모두) |
|---|---|---|
| **before (random init)** | "자주 등장하는 *관사·전치사·특수문자* (`the`, `a`, `,`, `.`, `of`)" | `broadband, ##eti, weights, smack, ##umen` / `bishop, sentencing, [unused174], chip, monitors` — **조각·미사용 토큰 등 난수** |
| **ours (2 epoch)** | "*방향성이 보이기 시작* — 일반 부사·형용사, 위키 어휘 일부" | **네 문장 모두** `the, ,, ., and, of` — **고빈도 기능어만** (실행에 따라 끝자리만 `of`/`in` 사이에서 흔들림) |

즉 본문이 before 자리에 써 놓은 설명이 실제로는 ours의 모습이고, before는 빈도와도 무관한 난수 토큰입니다. `logits 가 미세하게 흔들려 통계적 빈도 높은 토큰만 뽑힘`이라는 근거 서술도 성립하지 않습니다 — random init의 출력 분포는 코퍼스 빈도와 아무 관계가 없습니다.

같은 오해가 §6-3 도입부의 `top-5 가 *방향성은 맞지만 정답이 잘 안 보이는*` 에도 이어집니다. **입력 네 문장이 서로 달라도 top-5가 사실상 동일**하므로 "방향성"이라 부를 근거가 출력에 없습니다.

### (b) reference BERT 예시 토큰이 실제 출력과 다름 — Minor

본문: `위키 도메인은 정답이 top-1 — paris, zero 같은 자연스러운 답`

실제 출력:

| 입력 | `bert-base-uncased` top-5 |
|---|---|
| `The capital of France is [MASK].` | `paris, lille, lyon, marseille, tours` ✅ |
| `Water freezes at [MASK] degrees Celsius.` | `100, 60, 50, 30, 90` ❌ |

`zero`는 top-5에 없고, `bert-base-uncased`는 오히려 **상식적으로 틀린 답**을 냅니다. reference 모델은 고정 가중치라 이 출력은 **결정론적** 이고 2회 실행에서 동일했습니다 — 재실행으로 사라지지 않습니다.

## 기대 결과

- **(a)** before / ours 서술을 실제 출력에 맞게 교체. 특히 ours를 "방향성이 보인다"가 아니라 **"코퍼스 최상위 unigram만 재현하는 단계 = 문맥 미학습"** 으로 규정.
- **(b)** `zero` 예시 삭제. 나아가 `100`이 나오는 현상을 **"MLM은 사실이 아니라 *이 자리에 올 법한 토큰의 분포* 를 배운다"** 는 단서로 전환 (사전학습 모델을 지식 베이스로 오해하지 않게 하고, downstream fine-tune의 필요성으로 연결).

## 수정 제안

**(a)** 는 산문 교체로 해결됩니다. 학습 전(random init)에는 **빈도와도 문맥과도 무관한 난수 토큰** 이 나오고 학습 후에야 **고빈도 기능어** 가 자리를 잡는 것이 실제 순서이므로, 지금 before / ours 에 적힌 두 설명을 서로 맞바꾸면 됩니다. "방향성이 보이기 시작한다" 는 표현은 삭제하는 편이 맞습니다 — 입력 네 문장이 서로 달라도 top-5 가 같으므로 출력에 근거가 없습니다.

다만 문구만 바꾸면 학습 결과가 "그냥 실패" 로 읽힙니다. **학습 후 상태를 "unigram 분포까지 배운, MLM 학습이 반드시 거치는 중간 단계" 로 규정** 하는 것이 이 챕터의 결과를 가장 정확하게 설명합니다. 이를 뒷받침하려면 학습 코퍼스의 unigram cross-entropy 를 계산해 *균등 → unigram → 우리 모델* 사다리를 보여주는 짧은 셀을 세 모델 비교 직전에 두는 방안을 권고합니다 (수치는 위 「추가 확인」 참고). 추가 다운로드·학습 없이 이미 메모리에 있는 데이터만 쓰면 되고, 본편의 높은 perplexity 가 "짧은 데모 학습의 정직한 반영" 이라는 부록 #19 서술과도 맞물립니다.

**(b)** 는 `zero` 인용을 지우는 데서 끝내기보다, **reference BERT 가 `100` 을 내놓는다는 사실 자체를 교육 포인트로 전환** 하는 편이 낫습니다. MLM 이 배우는 것은 세상의 사실이 아니라 "이 자리에 올 법한 토큰의 분포" 이며, 그래서 downstream task 마다 fine-tune 이 필요하다는 흐름으로 자연스럽게 이어집니다.

아울러 세 모델을 가르는 축을 **"정답을 맞히는가" 에서 "입력이 바뀌면 답도 바뀌는가" 로 옮길 것** 을 제안합니다. 문맥을 쓰는지 여부를 직접 드러내는 축이라 설명이 정확해지고, 실행 편차에도 흔들리지 않습니다.

마지막으로 수정 문구에는 **특정 실행의 토큰을 인용하지 않는 것** 이 안전합니다. random init 출력은 매 실행 바뀌고, 학습 후 top-5 도 끝자리가 빈도 엇비슷한 토큰 사이에서 흔들립니다. `##…` 조각 토큰·`[unused…]` 슬롯 같은 **패턴** 으로 서술하고, "네 문장이 완전히 동일" 같은 단정은 피하는 편이 좋습니다.

## 재현 절차

README의 Ch 20 Colab 배지로 노트북을 열어 전체 실행 후, 다음 세 출력과 본문 서술을 대조합니다.

1. "🔬 학습 직전 baseline" 셀 출력의 `top-5 before pretraining`
2. §6-1 셀 출력의 `top-5 after pretraining`
3. §6-4 3-way 비교 표의 `ref (bert-base)` 행

정적 대조만으로도 확인 가능합니다.

```bash
python3 -c "
import json,re
nb=json.load(open('executed/20_en_bert_pretrain.ipynb'))
for c in nb['cells']:
    for o in c.get('outputs',[]):
        t=''.join(o.get('text',''))
        if 'top-5 before' in t or 'ref   (bert-base)' in t:
            print(re.sub(r'\x1b\[[0-9;]*m','',t))
"
```

## 로그

**학습 전 (random init)** — 2026-08-09 Colab T4 실행:

```
input: The capital of France is [MASK].
  top-5 before pretraining: ['bishop', 'sentencing', '[unused174]', 'chip', 'monitors']
input: Water freezes at [MASK] degrees Celsius.
  top-5 before pretraining: ['##rem', 'gestapo', '##dict', '[unused368]', 'sentencing']
```

**학습 후 (2 epoch)** — 입력이 달라도 답이 동일:

```
input: The capital of France is [MASK].
  top-5 after pretraining: ['the', ',', '.', 'and', 'of']
input: Water freezes at [MASK] degrees Celsius.
  top-5 after pretraining: ['the', ',', '.', 'and', 'of']
input: The food at this restaurant was absolutely [MASK].
  top-5 after pretraining: ['the', ',', '.', 'and', 'of']
input: I would [MASK] recommend this place.
  top-5 after pretraining: ['the', ',', '.', 'and', 'of']
```

**3-way 비교 표의 ref 행**:

```
input: Water freezes at [MASK] degrees Celsius.
  ref   (bert-base)      : 100, 60, 50, 30, 90
```

레포의 `executed/20_en_bert_pretrain.ipynb` 도 동일한 양상입니다 (before만 난수라 토큰이 다름: `broadband, ##eti, weights, smack, ##umen`).

## 추가 확인 — 왜 이런 출력이 나오는가 (수정 문구의 근거)

검토 중 임시 셀로 **학습 코퍼스의 unigram 분포**를 계산해 보면 현상이 정확히 설명됩니다 (같은 Colab T4 실행, 추가 학습 없음):

| 기준선 | eval loss (nats) | perplexity |
|---|---|---|
| 균등 분포 `ln V` (random init) | 10.3262 | 30,522 |
| **unigram — 문맥 없이 빈도만 아는 모델** | **7.2525** | **1,412** |
| 우리 작은 BERT (2 epoch, 실측) | **7.1259** | **1,244** |

모델이 unigram 기준선보다 **0.1266 nats (약 1.8%)** 나은 데 그칩니다. 그리고 코퍼스 최상위 unigram은 `the, ,, ., of, and` 로 **모델의 top-5와 사실상 동일**합니다.

즉 2 epoch이 새긴 것은 대부분 **빈도 통계**이고 문맥은 거의 학습되지 않았습니다. **입력이 바뀌어도 top-5가 안 바뀌는 것은 버그가 아니라 필연** 이며, 본편의 높은 perplexity가 "짧은 데모 학습의 정직한 반영"이라는 부록(#19) 서술과도 일관됩니다.

수정 시 이 기준선을 계산하는 셀을 함께 넣으면, 지금 "실패처럼 보이는 결과"가 **MLM 학습이 반드시 거치는 첫 계단**으로 읽히고 Ch 21의 downstream 성능(정확도 0.626)과도 연결됩니다.

## 전파 범위

- (a) `pages/20-en_bert_pretrain-practice.md:566` / `pages/20-en_bert_pretrain-anatomy.md:176,302,303`
- (b) `pages/20-en_bert_pretrain-anatomy.md:304`
- (a) `book/chapters/ch20_en_bert_pretrain.tex:1127,1582,1759,1761`
- (b) `book/chapters/ch20_en_bert_pretrain.tex:1763`

## 심각도

**Major** — (a)는 챕터의 핵심 관찰(사전학습 전·후 비교)에 대한 설명이 실제 출력과 반대이고, 독자가 바로 위 셀 출력만 봐도 모순을 확인할 수 있습니다. (b)는 단독으로는 Minor이나 같은 셀이라 함께 수정합니다. 실행은 정상 완주하므로 Blocker는 아닙니다.

## 수정 경로 주의

- **산문만 고치는 경우**: 재실행 없이 WikiDocs 재변환 + EPUB 재빌드만 필요합니다.
- **unigram 기준선 셀을 추가하는 경우**(권고): 코드가 바뀌므로 재실행 → `executed/` 갱신 → WikiDocs 재변환 → EPUB 재빌드가 필요하고 Ch 20의 WikiDocs·EPUB PASS가 무효화됩니다 (계획서 §8). 노트북 전체가 약 82초라 재실행 부담은 작습니다.
- 같은 재실행에 묶을 후보: `encoder body 자체는 약 2M` (실측 3,159,040) 표기 오류, `trainer.evaluate()` 중복 호출로 §6과 §6-2의 `eval_loss` 가 서로 다른 값으로 출력되는 건.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자
- 연결: 챕터 검수 이슈 #44

**Comment by fluentmin** [comment]
## 추가 검출 — 같은 재실행에 묶을 두 건 (c)(d)

본문 「수정 경로 주의」에서 후보로만 적었던 두 건을 근거와 함께 확정합니다. 둘 다 **Minor** 이며, (a)(b) 수정으로 어차피 재실행이 필요하므로 **같은 사이클에 묶어 처리**하는 것을 전제로 합니다.

### (c) 작은 BERT 의 파라미터 수치가 실제 출력과 다름

본문은 파라미터 구성을 `임베딩 테이블이 절반 이상 … encoder body 자체는 약 2M` 이라고 서술합니다. 그러나 같은 셀이 바로 위에서 출력하는 실측값은 다릅니다.

| 구성 | 본문 서술 | 실제 출력 |
|---|---|---|
| embeddings | "절반 이상" | **7,847,424 (70.7%)** |
| encoder (4 layer) | "약 2M" | **3,159,040 (28.5%)** |
| MLM head | — | 96,826 (0.9%) |

`encoder body 약 2M` 은 실제의 **약 1.6배 과소** 입니다. hidden 256 · 4층이면 층당 attention 4×256×256 ≈ 262K + FFN 256×1024×2 ≈ 524K 로 약 790K, 4층이면 약 3.16M 이므로 출력 쪽이 맞습니다. "절반 이상" 도 틀린 말은 아니나 실제 70.7% 를 과소 표현해, *"vocab 큰데 모델이 작은 셋업"* 이라는 이 문단의 논지를 오히려 약하게 만듭니다.

모델 config 가 고정이라 이 값은 **결정론적** 입니다 — 실행 3회에서 모두 `7,847,424 / 3,159,040 / 96,826` 로 동일했고, 재실행으로 사라지지 않습니다.

### (d) 같은 모델·같은 평가셋인데 `eval_loss` 가 두 값으로 출력됨

학습이 끝난 뒤 `trainer.evaluate()` 가 **두 번** 호출됩니다 — §6 평가 절에서 한 번, §6-1 사전·사후 비교에서 또 한 번. MLM 은 collator 가 매번 무작위로 마스킹하므로 두 호출의 결과가 미세하게 다르고, 그 두 값이 한 노트북 안에 나란히 실립니다.

| 실행 | §6 출력 | §6-1 이후 출력 |
|---|---|---|
| 2026-06-22 (기존 실행본) | 7.1202 (ppl 1,236.73) | 7.1278 (ppl 1,246.14) |
| 2026-08-09 ① | 7.1248 (ppl 1,242.40) | 7.1259 (ppl 1,243.73) |
| 2026-08-09 ③ (현재 브랜치) | 7.1339 (ppl 1,253.81) | 7.1347 (ppl 1,254.79) |

세 번 모두 재현됩니다. 독자 입장에서는 **같은 모델의 같은 지표가 두 숫자로 보이므로 오탈자로 읽히기 쉽습니다.** 실제로는 버그가 아니라 MLM 평가의 확률적 성질인데, 노트북에 그 설명이 없어 정보가 되지 못하고 혼선만 남습니다. 덤으로 불필요한 평가 1회분의 시간도 씁니다.

## 수정 제안

(c) 는 문장의 수치를 실측에 맞추면 됩니다. 다만 단순 교체에 그치지 말고 **임베딩이 70% 를 넘는다는 사실을 그대로 드러내는 편** 이 좋겠습니다 — 이 문단의 목적이 "vocab 은 큰데 본체는 작다" 를 보이는 것이므로, 정확한 비율이 논지를 더 선명하게 만듭니다.

(d) 는 **평가를 한 번만 수행하고 그 값을 사후 비교에서 재사용** 하는 방향을 제안합니다. 두 숫자가 사라지고 실행 시간도 조금 줄어듭니다. 나아가 **왜 다시 부르면 값이 달라지는지 한 줄로 밝혀 두면** 지금의 혼선거리가 오히려 MLM 평가의 성질을 알려주는 지점이 됩니다 — 마스킹이 매 호출 무작위라 같은 모델도 값이 흔들리며, 그래서 한 번 잰 값을 고정해 쓴다는 설명입니다.

## 전파 범위

- (c) `pages/20-en_bert_pretrain-practice.md:364`, `book/chapters/ch20_en_bert_pretrain.tex:818`
- (d) `pages/20-en_bert_pretrain-anatomy.md:36,70` (두 `trainer.evaluate()` 호출과 각각의 출력)

## 심각도

둘 다 **Minor** — 실행은 정상이고 챕터의 논지를 뒤집지 않습니다. 다만 (c) 는 코드가 바로 위에서 반박하는 수치라 그대로 두면 눈에 띄고, (d) 는 오탈자로 오해받기 쉽습니다. (a)(b) 재실행에 함께 태우면 추가 비용이 사실상 없습니다.
